src/board/system76/common/flash/flash.mk

```diff
@@ -0,0 +1,56 @@
+# Set flash ROM parameters
+FLASH_OFFSET=2048
+FLASH_SIZE=1024
+CFLAGS+=-DFLASH_OFFSET=$(FLASH_OFFSET) -DFLASH_SIZE=$(FLASH_SIZE)
+
+# Copy parameters to use when compiling flash ROM
+FLASH_INCLUDE=$(INCLUDE)
+FLASH_CFLAGS=$(CFLAGS)
+
+# Include flash source.
+FLASH_DIR=$(SYSTEM76_COMMON_DIR)/flash
+# Note: main.c *must* be first to ensure that flash_start is at the correct address
+FLASH_SRC=$(FLASH_DIR)/main.c
+FLASH_INCLUDE+=$(wildcard $(FLASH_DIR)/include/flash/*.h) $(FLASH_DIR)/flash.mk
+FLASH_CFLAGS+=-I$(FLASH_DIR)/include -D__FLASH__
+
+# Add minimal source from other directories
+FLASH_SRC+=
+
+FLASH_BUILD=$(BUILD)/flash
+FLASH_OBJ=$(patsubst src/%.c,$(FLASH_BUILD)/%.rel,$(FLASH_SRC))
+FLASH_CC=\
+	sdcc \
+	-mmcs51 \
+	--model-large \
```

Comment:
The large model will potentially use data space also used by the main ROM. Is there a reason to not use --model-small like the scratch ROM?

Reply:
From the sdcc manual:
When the small model is used variables without an explicitly specified intrinsic named address space are allocated in the internal ram.
When the medium, large or huge model is used all variables declared without specifying an intrinsic named
address space will be allocated into the external ram, this includes all parameters and local variables (for nonreentrant functions). Medium model uses pdata and large and huge models use xdata.

Effectively, -small causes the main data segment to be mangled, while large causes external data to be mangled.
In reality, it shouldn't matter so long as reentrant is used, and this ensures that only the stack is used.

Unfortunately, the main reason for one over the other is due to the runtime libraries that get pulled in. When -small is used, _gptrput gets pulled in with PARAM_2 in OSEG (data). When using -large, this is placed in XSEG instead (xdata). Note that ideally this would be reentrant as well, however the sdcc doesn't have a reentrant version of _gptrput.

The xdata location that is allocated for _gptrput is unused according to ec.map and flash.map, while the location in data was already used.